src/Servers/Kestrel/Core/src/KestrelServerOptions.cs

```diff
@@ -38,7 +38,20 @@ public class KestrelServerOptions
     // The following two lists configure the endpoints that Kestrel should listen to. If both lists are empty, the "urls" config setting (e.g. UseUrls) is used.
     internal List<ListenOptions> CodeBackedListenOptions { get; } = new List<ListenOptions>();
     internal List<ListenOptions> ConfigurationBackedListenOptions { get; } = new List<ListenOptions>();
-    internal IEnumerable<ListenOptions> ListenOptions => CodeBackedListenOptions.Concat(ConfigurationBackedListenOptions);
+
+    internal ListenOptions[] GetListenOptions()
```

Comment:
Would it really be that bad to leave it as a property? It's not that expensive, and might be easier than tracking down all the tests.

Reply:
I can leave it as a property, but I really dislike allocating in a property getter.

Reply:
Question: are the two lists mutated much at the moment? Thery *could* be stored in a single list or array, with an int that marks the split between the two chunks. Then the combined list is "ignore the split", and the other two lists are just separate Alice's of the single chunk - ideally returning either RwadOnlyMemory-ListenOptions, ReadOnlySpan-ListenOptions, or ArraySegment-ListenOptions. If it is easiest to use a list when building, theres a CollectionsMarshal API to sneak inside when needed

Reply:
From a quick investigation, for a `dotnet new api` template app, neither list is mutated.

The way `CodeBackedListenOptions` gets mutated is if someone adds endpoints in code, ex: `builder.WebHost.ConfigureKestrel(o => o.ListenLocalhost(5556));`.

The way `ConfigurationBackedListenOptions` get mutated is if someone adds endpoints through configuration, ex: `builder.Configuration.AddInMemoryCollection(new Dictionary<string, string?>() { { "Kestrel:EndPoints:Http:Url", "http://localhost:5555" } });`.

They are kept separate in order to correctly "reload" on configuration changes. Kestrel knows which ones were added by the previous configuration, and can remove any endpoints that are now removed.

I think your idea might be a good idea, but I think it is out of scope for this PR. Feel free to open a separate PR for it.